### PR TITLE
[KYUUBI #3893] [BUG] Fix spark extension: UnspecifiedDistribution does not have default partitioning.

### LINKED ISSUE #3893: [Bug] kyuubi spark extension throws java.lang.IllegalStateException: UnspecifiedDistribution does not have default partitioning.
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the bug

kyuubi extension rule 'InsertShuffleNodeBeforeJoin' seems not work at spark 3.3.1. When insert shuffle before skewed join, it throws java.lang.IllegalStateException: UnspecifiedDistribution does not have default partitioning.

After apache/spark#32816 ,OptimizeSkewedJoin was moved from queryStageOptimizerRules to queryStagePreparationRules.  So OptimizeSkewedJoin will be executed **before** InsertShuffleNodeBeforeJoin.  

After OptimizeSkewedJoin added a **AQEShuffleRead** node before SortExec,  the plan looks like 

``` CollectLimit 21
 +- Project [cast(key2#8L as string) AS key2#18, cast(value2#9L as string) AS value2#19]
    +- SortMergeJoin(skew=true) [key1#2L], [key2#8L], Inner
       :- Sort [key1#2L ASC NULLS FIRST], false, 0
       :  +- AQEShuffleRead
       :     +- ShuffleQueryStage 2
       :        +- Exchange hashpartitioning(key1#2L, 100), ENSURE_REQUIREMENTS, [plan_id=119]
       :           +- *(3) HashAggregate(keys=[key1#2L], functions=[], output=[key1#2L])
       :              +- ShuffleQueryStage 0
       :                 +- Exchange hashpartitioning(key1#2L, 100), ENSURE_REQUIREMENTS, [plan_id=67]
       :                    +- *(1) HashAggregate(keys=[key1#2L], functions=[], output=[key1#2L])
       :                       +- *(1) Project [CASE WHEN (id#0L < 25001) THEN 2499 WHEN (id#0L >= 75000) THEN 10000 ELSE id#0L END AS key1#2L]
       :                          +- *(1) Range (0, 102020, step=1, splits=8)
       +- Sort [key2#8L ASC NULLS FIRST], false, 0
          +- AQEShuffleRead skewed
             +- ShuffleQueryStage 1
                +- Exchange hashpartitioning(key2#8L, 100), ENSURE_REQUIREMENTS, [plan_id=81]
                   +- *(2) Project [CASE WHEN (id#6L < 5650) THEN 4999 WHEN (id#6L >= 10000) THEN 10001 ELSE id#6L END AS key2#8L, id#6L AS value2#9L]
                      +- *(2) Range (0, 20000000, step=1, splits=7)
```
then InsertShuffleNodeBeforeJoin throws exception:
``` java
Exception in thread "main" java.lang.IllegalStateException: UnspecifiedDistribution does not have default partitioning.
	at org.apache.spark.sql.catalyst.plans.physical.UnspecifiedDistribution$.createPartitioning(partitioning.scala:56)
	at org.apache.kyuubi.sql.InsertShuffleNodeBeforeJoin$.org$apache$kyuubi$sql$InsertShuffleNodeBeforeJoin$$withShuffleExec(KyuubiQueryStagePreparation.scala:91)
	at org.apache.kyuubi.sql.InsertShuffleNodeBeforeJoin$.org$apache$kyuubi$sql$InsertShuffleNodeBeforeJoin$$checkAndInsertShuffle(KyuubiQueryStagePreparation.scala:85)
	at org.apache.kyuubi.sql.InsertShuffleNodeBeforeJoin$$anonfun$insertShuffleBeforeJoin$1.applyOrElse(KyuubiQueryStagePreparation.scala:57)
	at org.apache.kyuubi.sql.InsertShuffleNodeBeforeJoin$$anonfun$insertShuffleBeforeJoin$1.applyOrElse(KyuubiQueryStagePreparation.scala:55)
```

Our fix:
``` scala
  private def checkAndInsertShuffle(
      distribution: Distribution,
      child: SparkPlan): SparkPlan = child match {
    case SortExec(_, _, _: Exchange, _) =>
      child
    case SortExec(_, _, _: QueryStageExec, _) =>
      child

    // add a pattern to filter this situation
    case SortExec(_, _, _: AQEShuffleReadExec, _) =>
      child

    case sort @ SortExec(_, _, agg: BaseAggregateExec, _) =>
      sort.withNewChildren(withShuffleExec(distribution, agg) :: Nil)
    case _ =>
      withShuffleExec(distribution, child)
  }
```

easy way to reproduce:
``` scala
    val spark = SparkSession.builder()
      .config("spark.sql.autoBroadcastJoinThreshold", "-1")
      .config("spark.sql.extensions", "org.apache.kyuubi.sql.KyuubiSparkSQLExtension")
      .config("spark.sql.optimizer.forceShuffleBeforeJoin.enabled", "true")
      .config("spark.sql.shuffle.partitions", "100")
      .config("spark.sql.adaptive.skewJoin.skewedPartitionThresholdInBytes", "80")
      .config("spark.sql.adaptive.advisoryPartitionSizeInBytes", "80")
      .config("spark.sql.adaptive.coalescePartitions.minPartitionNum", "20")
      .config("spark.sql.adaptive.coalescePartitions.minPartitionSize", "10")
      .config("spark.sql.planChangeLog.level", "info")
      .master("local[*]").getOrCreate()
    import spark.implicits._
    spark
      .range(0, 102020, 1, 8)
      .select(
        when('id < 25001, 2499)
          .when('id >= 75000, 10000)
          .otherwise('id).as("key1"),
        'id as "value1")
      .createOrReplaceTempView("skewData1")
    spark
      .range(0, 20000000, 1, 7)
      .select(
        when('id < 5650, 4999)
          .when('id >= 10000, 10001)
          .otherwise('id).as("key2"),
        'id as "value2")
      .createOrReplaceTempView("skewData2")
    val join =
      s"""
         | SELECT key2,value2 FROM
         | (SELECT  key1, count(1) as cnt FROM skewData1 group by key1)
         | JOIN (select * from skewData2) ON key1 = key2
         |""".stripMargin
    spark.sql(join).show()
```
pom
``` xml
        <dependency>
            <groupId>org.apache.spark</groupId>
            <artifactId>spark-hive_2.12</artifactId>
            <version>3.3.1</version>
        </dependency>
        <dependency>
            <groupId>org.apache.kyuubi</groupId>
            <artifactId>kyuubi-extension-spark-common_2.12</artifactId>
            <version>1.6.1-incubating</version>
        </dependency>
        <dependency>
            <groupId>org.apache.kyuubi</groupId>
            <artifactId>kyuubi-extension-spark-3-3_2.12</artifactId>
            <version>1.6.1-incubating</version>
        </dependency>
```

### Affects Version(s)

master

### Kyuubi Server Log Output

_No response_

### Kyuubi Engine Log Output

_No response_

### Kyuubi Server Configurations

_No response_

### Kyuubi Engine Configurations

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes. I can submit a PR independently to fix.
- [X] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to fix.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by FireFoxAhri**
### _Why are the changes needed?_

1. to fix #3893 

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
```
Error:  /home/runner/work/incubator-kyuubi/incubator-kyuubi/extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:25: object AQEShuffleReadExec is not a member of package org.apache.spark.sql.execution.adaptive
Error:  /home/runner/work/incubator-kyuubi/incubator-kyuubi/extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:82: not found: type AQEShuffleReadExec
Error: [ERROR] two errors found
Error:  Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project kyuubi-extension-spark-common_2.12: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
```

**Comment by FireFoxAhri** [comment]
> https://github.com/apache/incubator-kyuubi/actions/runs/3608352426/jobs/6081140138
> 
> ```
> Error:  /home/runner/work/incubator-kyuubi/incubator-kyuubi/extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:25: object AQEShuffleReadExec is not a member of package org.apache.spark.sql.execution.adaptive
> Error:  /home/runner/work/incubator-kyuubi/incubator-kyuubi/extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:82: not found: type AQEShuffleReadExec
> Error: [ERROR] two errors found
> Error:  Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile (scala-compile-first) on project kyuubi-extension-spark-common_2.12: Execution scala-compile-first of goal net.alchim31.maven:scala-maven-plugin:4.6.1:compile failed: Compilation failed: InterfaceCompileFailed -> [Help 1]
> ```

Looks like 3.1 doesn't have AQEShuffleReadExec, and the problem currently only appears in 3.3.x. Should I separate InsertShuffleNodeBeforeJoin to different module and only change spark 3.3 module

File: extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala [hunk 1]
```diff
@@ -79,6 +79,8 @@ object InsertShuffleNodeBeforeJoin extends Rule[SparkPlan] {
       child
     case SortExec(_, _, _: QueryStageExec, _) =>
       child
+    case SortExec(_, _, _: AQEShuffleReadExec, _) =>
+      child
```
Comment (turboFei) [inline extensions/spark/kyuubi-extension-spark-common/src/main/scala/org/apache/kyuubi/sql/KyuubiQueryStagePreparation.scala:83]: > Looks like 3.1 doesn't have AQEShuffleReadExec, and the problem currently only appears in 3.3.x. Should I separate InsertShuffleNodeBeforeJoin to different module and only change spark 3.3 module

Maybe you can check the className here?
```
case SortExec(_, _, e, _) if  e.getClass.getName == "org.apache.spark.sql.execution.adaptive.AQEShuffleReadExec" =>
  child
```

**Comment by ulysses-you** [comment]
thank you @FireFoxAhri for the work. The reason is since Spark3.3, optimize skew join moved to preparation phase. Then the rule ordering conflicts.

**Comment by ulysses-you** [comment]
I think a better way is we can skip adding shuffle node if the smj or shj is skewed.

**Comment by FireFoxAhri** [comment]
> I think a better way is we can skip adding shuffle node if the smj or shj is skewed.

Like this? I also think it's better. 
``` scala
  private def insertShuffleBeforeJoin(plan: SparkPlan): SparkPlan = plan transformUp {
    case smj @ SortMergeJoinExec(_, _, _, _, l, r, isSkewJoin) if !isSkewJoin =>
    case shj: ShuffledHashJoinExec if !shj.isSkewJoin =>
```
but spark 3.1.x shj doesn't have isSkewJoin

**Comment by ulysses-you** [comment]
I'm fine to separate this rule, and change with Spark 3.3

**Comment by FireFoxAhri** [comment]
Copy & paste the code, but IDEA auto formatted when paste.

**Pull request closed without merging**